Docs: Okta integration docs overhaul

kopiczko opened the pull request:
Notes to reviewers:

- Note on how the review could be approached https://github.com/gravitational/teleport/pull/52821#issuecomment-2704201840
- I wanted to sort pages in logical order in the left sidebar. I tried `sidebar_position` from [docosaurus docs](https://docusaurus.io/docs/api/plugins/@docusaurus/plugin-content-docs#markdown-front-matter) but it doesn't seem to work. I also tried adding numbers on the front of the files, but that didn't work either.
- ~~Some file titles/slugs don't look appropriate. This is because I was reusing old paths in case some users have bookmarked the pages. I didn't manage to find a way to create URL aliases.~~ I found about `docs/config.json` file.

| Starting path for reading in Amplify `/enroll-resources/application-access/okta/` |
|---|

kopiczko committed:
Docs: Okta integration docs overhaul

ptgott reviewed (commented):
Hi @kopiczko, would you mind including a summary of changes in the PR description? Since this is an overhaul, and there are 54 files being changed, it's a little difficult to predict how much time I'll need to review this. If this PR rewrites multiple guides, I would consider multiple PRs to make it easier to wave through less-feedback-intensive work.

Also, what is your desired sidebar structure? We control the [sidebar sorting logic](https://github.com/gravitational/docs-website/blob/main/server/sidebar-order.ts#L53-L105), so we can make any adjustments. I think the Docusaurus `sidebar_position` attribute may get lost in the our sidebar sorter, but we can fix that.

kopiczko commented:
Hi @ptgott, bulk of those are screenshots, so if you collapse the img directory there aren't anywhere near 54 files (see the screenshot below). 

When reviewing I'd suggest starting with the file `docs/pages/enroll-resources/application-access/okta/integration-enrollment.mdx` highlighted on the screenshot. Or in Amplify with `/enroll-resources/application-access/okta/` URL path.

In general the whole content of `docs/pages/enroll-resources/application-access/okta/` was heavily changed, but `integration-enrollment.mdx` is the starting point and links to every other file. The rest is mostly fixes to linked files after the renames.

<img width="357" alt="Screenshot 2025-03-06 at 15 27 38" src="https://github.com/user-attachments/assets/20acbf47-3d40-4ef1-a0dd-4c9423018460" />

ptgott commented:
@kopiczko Re: the sidebar, if we want to make all guides in this section part of a series, we should make it explicit in the `title` or `sidebar_label` of each guide, e.g., by adding "Part n:" to the beginning of each guide's `sidebar_title` (that'll also configure the ordering, which happens alphabetically unless the title includes "Getting Started" or "Introduction). In nearly all subsections of the Enroll Resources section, sidebar titles don't belong to a series, so we don't set an expectation with the reader that the sidebar is organized in the order we want them to read the pages.

I'll make time to give this a look! I might make some suggestions for the information architecture. For example, it's not clear to me from the [category index page](https://kopiczko-okta-integration-docs-overhaul.d3pp5qlev8mo18.amplifyapp.com/enroll-resources/application-access/okta/) that the guides need to be read in order, or why the two Appendix pages are called "Appendix", and aren't additional how-to guides.

kiosion commented on docs/pages/enroll-resources/application-access/okta/app-and-group-sync.mdx:
```diff
@@ -0,0 +1,111 @@
+When first synchronizing an Access List, the owners of the Access List will be set to the default
+owners configured during the initial Okta integration enrollment, and the initial review date will
+be set 6 months from the configured date. These fields are modifiable, as well as the owner and
+membership requirements. However, you will not be able to change grants, as these belong to the
```
I'm not sure this is mentioned in the UI currently – do we want to?

kiosion reviewed (commented):
Some nitpicks on wording, but mostly looks good to me

zmb3 commented:
Do we absolutely need 30+ new screenshots? That's an awful lot of work to keep up to date every time the UI changes, and it bloats the repo further (making clones take longer, etc).

kopiczko commented:
> Do we absolutely need 30+ new screenshots? That's an awful lot of work to keep up to date every time the UI changes, and it bloats the repo further (making clones take longer, etc).

We could probably get away with less, but the question is if it's worth it reducing it by let's say half and reducing the quality of the docs significantly. It's likely that the Okta enrolment flow is one of the most complicated setups in Teleport as of right now so it requires quite a lot of explanations.

I have a feeling if we reduce the number of screenshots then the amount of text will increase and I'm not sure if this is any easier to maintain. It will be can difficult to spot something is outdated in the prose. Also it's way less approachable.

Having said that I'm open to suggestions.

ptgott commented on docs/pages/enroll-resources/application-access/okta/app-and-group-sync.mdx:
```diff
@@ -0,0 +1,111 @@
+---
+title: Okta apps and groups sync
```
We use title case for titles. I would give the [Style guide](https://www.notion.so/goteleport/Style-guide-160fdd3830be80a79e70f08df2df2a0a) a read when you get a chance.

ptgott commented on docs/pages/enroll-resources/application-access/okta/app-and-group-sync.mdx:
```diff
@@ -0,0 +1,111 @@
+---
+title: Okta apps and groups sync
+description: Details on Okta integration apps and groups synchronization as Teleport Access Lists
```
```suggestion
description: Provides details on using the Okta integration to synchronization applications and groups as Teleport Access Lists.
```
- Fixing grammar
- Beginning the description with a verb per the style guide

ptgott commented on docs/pages/enroll-resources/application-access/okta/app-and-group-sync.mdx:
```diff
@@ -0,0 +1,111 @@
+description: Details on Okta integration apps and groups synchronization as Teleport Access Lists
+---
+
+Okta has its own permissions system that doesn't directly map into Teleport. This
+includes permissions granted by Okta groups along with the assignment of users to individual
+applications within Okta. To model this within Teleport, an administrator would typically need
+to carefully craft a labeling system to attach to various Okta apps and groups and a set of
+roles to go along with them.
+
+With the synchronization of Okta apps and groups ass Teleport Access List feature in the Okta
+integration, this work is performed for you. We'll discuss the details of how this works.
```
Is this an architectural guide? When would a user read this guide?

ptgott commented on docs/pages/enroll-resources/application-access/okta/app-and-group-sync.mdx:
```diff
@@ -0,0 +1,111 @@
+(optional), *Level 3. User sync* and *Level 4. App and Group sync* enrollment in the [Okta
+integration setup guide](./integration-enrollment.mdx).
+
+## Synchronizing Access Lists from Okta groups and applications
```
If this is an architecture guide, we should move it to `reference/architecture`. An alternative would be to move the general architecture information to the section index page at `enroll-resources/application-access/okta/okta.mdx`. This would provide a conceptual overview of the feature so a user has a mental model to work with when navigating through the different pages.

ptgott commented on docs/pages/enroll-resources/application-access/okta/app-and-group-sync.mdx:
```diff
@@ -0,0 +1,111 @@
+With the synchronization of Okta apps and groups ass Teleport Access List feature in the Okta
+integration, this work is performed for you. We'll discuss the details of how this works.
+
+## Enrollment
```
I'm not sure why the first H2 of this page is a description of other pages. I think this guide would benefit from a better-defined purpose so it doesn't overlap as much with other content.

ptgott commented on docs/pages/enroll-resources/application-access/okta/app-and-group-sync.mdx:
```diff
@@ -0,0 +1,111 @@
+When synchronizing from Okta to Teleport, the synchronizer handles the flattened structure from Okta
+by attempting to map it back to the Teleport hierarchy:
+
+- **Comparing with Teleport Hierarchy**: The flattened list of members from Okta is compared against the existing
+Access List hierarchy in Teleport.
+- **Adding New Members**:
+  - **New Users**: If there are new users added in Okta that are not present in the Teleport Access List hierarchy,
+    these users are added as members at the root-level Access List in Teleport.
+  - **Maintaining Hierarchy**: This approach maintains the hierarchical structure within Teleport while ensuring
+    that access changes made in Okta are reflected appropriately.
```
I think the dense outline structure here is a little distracting I would use paragraphs instead.

ptgott commented on docs/pages/enroll-resources/application-access/okta/appendix-a-okta-saml-app.mdx:
```diff
@@ -0,0 +1,17 @@
+---
+title: Appendix A. Okta SAML app
```
- Why are we calling this an appendix?
- We should edit the title and description per the style guide.

ptgott commented on docs/pages/enroll-resources/application-access/okta/appendix-a-okta-saml-app.mdx:
```diff
@@ -0,0 +1,17 @@
+sidebar_position: 6
+---
+
+This page will guide you how to create a new Okta SAML 2.0 app for *Level 1. SSO* integration. It
```
I would find a way to remove the appendices, since they add unnecessary complexity to the docs.

Is there a way to inline this content as a partial within the SSO integration guide, SCIM guide, and user sync guide? Otherwise, it muddies the structure of the Okta Integration with Application Access section, and a user who discovers this guide needs to figure out for themselves how it relates to the other guides mentioned in the introductory paragraph.

ptgott reviewed (commented):
Leaving some initial feedback. I am going to read `integration-enrollment.mdx` and `user-sync.mdx` next.

ptgott commented on docs/pages/enroll-resources/application-access/okta/user-sync.mdx:
```diff
@@ -0,0 +1,102 @@
+---
+title: Okta user sync
```
We should edit the title and description per the style guide

ptgott commented on docs/pages/enroll-resources/application-access/okta/user-sync.mdx:
```diff
@@ -0,0 +1,102 @@
+description: Details on Okta integration user synchronization
+---
+
+Teleport uses two complimentary methods for synchronizing Teleport resources with
```
```suggestion
Teleport uses two complementary methods for synchronizing Teleport resources with
```

ptgott commented on docs/pages/enroll-resources/application-access/okta/user-sync.mdx:
```diff
@@ -0,0 +1,102 @@
+an upstream Okta Organization:
+
+- the Okta integration user sync process
+- provisioning via SCIM
```
Would a reader know what SCIM is, or should we expand the abbreviation here?

ptgott commented on docs/pages/enroll-resources/application-access/okta/okta.mdx:
```diff
@@ -4,9 +4,12 @@ description: Guides for using Teleport Okta integration.
 ---
 
-Configure Teleport to import and grant access to Okta applications and user groups.
+Configure Teleport to import Okta users and grant access to Okta applications and user groups as
```
Also, this section contains a combination of how-to guides and conceptual guides, so this would be a good opportunity to clarify when a user would read each kind of guide.

ptgott commented on docs/pages/enroll-resources/application-access/okta/user-sync.mdx:
```diff
@@ -0,0 +1,102 @@
+- the Okta integration user sync process
+- provisioning via SCIM
+
+Okta user sync works in tandem with [SCIM integration](./scim-integration.mdx#how-it-works). Unlike
```
```suggestion
Okta user sync works in tandem with the [Teleport SCIM integration](./scim-integration.mdx#how-it-works). Unlike
```
- Grammar
- Clarifying that we're referring to the Teleport SCIM integration

ptgott commented on docs/pages/enroll-resources/application-access/okta/user-sync.mdx:
```diff
@@ -0,0 +1,102 @@
+- provisioning via SCIM
+
+Okta user sync works in tandem with [SCIM integration](./scim-integration.mdx#how-it-works). Unlike
+SCIM integration, it is initiated from the Teleport side, and therefore requires API access to Okta
+and it is slower than SCIM, but it provides following benefits:
```
```suggestion
SCIM integration, it is initiated from the Teleport side, and therefore requires API access to Okta. It is slower than SCIM, but it provides following benefits:
```

ptgott commented on docs/pages/enroll-resources/application-access/okta/user-sync.mdx:
```diff
@@ -0,0 +1,102 @@
+SCIM integration, it is initiated from the Teleport side, and therefore requires API access to Okta
+and it is slower than SCIM, but it provides following benefits:
+
+- robustness:
+  - it does retries, it will fetch the users if for whatever reason SCIM request doesn't reach
+    the Teleport cluster
+  - it will fetch users that are assigned to the SSO app before SCIM was enabled
+- provides ability to synchronize users for the private Teleport clusters which are not accessible
+  over internet
+- unlocks enrollment of [synchronizing Okta apps and groups](./app-and-group-sync.mdx) as
+  Teleport Access Lists 
```
I would provide a couple of quick reasons instead of a detailed, multi-level outline.

ptgott commented on docs/pages/enroll-resources/application-access/okta/integration-enrollment.mdx:
I think we need to rework the structure of this section. Right now, there's one rather complicated how-to guide supplemented by several conceptual guides. The conceptual guides tell the reader to follow specific sections of the how-to guide. For me, this structure is difficult to navigate.

Instead, I would either:

a. Split up the how-to guide so that each level is a separate guide. Then shorten each conceptual guide and make it the "How it works" section of each how-to guide.
b. If (a) is not feasible, create a subsection of the `application-access/okta` section for each level. That way, each subsection can include a conceptual guide and a how-to guide.

If one level requires another level, we can mention the requirements in the Prerequisites section of the how-to guide for that level.

kopiczko replied:
> If one level requires another level, we can mention the requirements in the Prerequisites section of the how-to guide for that level.

All levels are dependent on all previous levels, with an exeception that Level 2. SCIM can be skipped. Also when you enroll the integration you can go trough all the levels.

When it comes to shortening the conceptual guides - I really don't know how to do that so they are significantly shorter. They are also mostly a copied text/moved files.

So if we really want to split the how-to guide I think we are left only with the option b.

Do you envision something like below?

```
Okta Integration with Application Access
    - Deleting Okta integration     // really strange it will go at the top IMO
    > Level 1. SSO
        - How to enroll
        - How it works
        - Appendix A. ...
        - Appendix B. ...
    > Level 2. SCIM
        - How to enroll
        - How it works
    > Level 3. User sync
        - How to enroll
        - How it works
    > Level 4. App and group sync
        - How to enroll
        - How it works
    - Okta integration status page   // not really visible
```

r0mant replied:
I wrote my [review comment](https://github.com/gravitational/teleport/pull/52821#pullrequestreview-2718576966) below and only afterwards saw that you folks have already discussed literally the same thing here. TLDR I agree with this proposal to split things up, I think we're on the same page and the proposed structure looks good to me 👍

r0mant reviewed (commented):
@kopiczko @ptgott I read through the [rendered version](https://kopiczko-okta-integration-docs-overhaul.d3pp5qlev8mo18.amplifyapp.com/enroll-resources/application-access/okta/integration-enrollment/#level-3-user-sync) of the guide and have some thoughts.

Overall, I like these changes and I think that having a companion guide like this to the guided web UI based flow is useful. I think the guide is well structured content-wise however I do agree that it's quite unwieldy. Things that I think are suboptimal:

- The guide is *huge* - shoving all 4 levels on the same page can be overwhelming for the reader.
- The "appendix a/b" thing disrupts the reader flow where I have to open another tab to learn how to create an Okta app for example and then go back.

With that in mind, I propose the following structural changes which shouldn't require rewriting anything but just shuffling some things around:

- Split each "level" into a separate guide. If level 2+ guides refer to others as prerequisites, that's fine.
- Get rid of "appendices" and instead include the content in the guides themselves, maybe in collapsible sections - if each page is smaller, then it wouldn't be as overwhelming.

WDYGT?

ptgott commented:
> With that in mind, I propose the following structural changes which shouldn't require rewriting anything but just shuffling some things around:

> Split each "level" into a separate guide. If level 2+ guides refer to others as prerequisites, that's fine.
Get rid of "appendices" and instead include the content in the guides themselves, maybe in collapsible sections - if each page is smaller, then it wouldn't be as overwhelming.
WDYGT?

@r0mant @kopiczko I am working on a branch based against this one that shuffles around the text into one guide for each level. I am planning to open a PR either today or tomorrow that we can review and merge into this one.

ptgott commented:
I've opened a PR here: 
- https://github.com/gravitational/teleport/pull/53510

Note that the PR includes some unrelated files, such as the feature matrix and `sidebar.json`. This is to get the branch to render on the dev server. Once we rebase this branch against `master` and rebase my branch against this one, the unrelated files should disappear from the diff.

kopiczko commented:
@ptgott I rebased my branch. Is it possible to see your branch rendered? I looked at the raw files but I'm not sure how it will read rendered given how many screenshots are stripped down.

Let's please proceed reasonably quickly. If we branch off too much that will never end. I'm happy to apply the wording suggestions and split the docs as proposed here https://github.com/gravitational/teleport/pull/52821#discussion_r2006355015. We can always iterate and currently we have very outdated docs.

ptgott commented:
@kopiczko It looks like our Amplify setup is having issues rendering a preview site for a stacked PR. I'll base my PR against `master` instead. The changes I've suggested implement the proposed restructuring, fix a number of grammatical issues, and edit the text to adhere to the documentation style guide.
